Leverage native heap delete mechanism for persistent tables.

ashwinstar opened the pull request:
Persistent tables historically implemented different mechanism to maintain free tuples by using on-disk freelist chain of itself. It proved extremely hard to maintain this free-list and required lot of supporting code to maintain and validate its integrity. This PR leverages the native heap delete and vacuum framework to manage tuple deletion for persistent tables as well. Required corresponding changes in crash recovery and related to places to handle the same. Tried to break-down the patch in easy to review short independent code changes.

asimrp commented on src/include/storage/bufpage.h:
```diff
@@ -371,7 +371,6 @@ typedef PageHeaderData *PageHeader;
 )
 #define PageSetPrunable(page, xid) \
 do { \
-	Assert(TransactionIdIsNormal(xid)); \
```
If removing the assert seems too radical, it can be weakened slightly: 
```
   Assert(TransactionIdIsNormal(xid) || xid == FrozenTransactionId)
```

asimrp commented on src/backend/cdb/cdbpersistentstore.c:
```diff
@@ -1566,33 +839,17 @@ void PersistentStore_AddTuple(
+							values,
+							flushToXLog,
+							persistentTid);
+	Assert(persistentTid->ip_posid != 0);
```
Can we please use `ItemPointerIsValid()` macro here?

asimrp commented on src/backend/cdb/cdbpersistentstore.c:
```diff
@@ -1637,55 +892,16 @@ void PersistentStore_FreeTuple(
-		
-	frozen_heap_inplace_update(persistentRel, persistentTuple);
-
+	simple_heap_delete_xid(persistentRel, persistentTid, FrozenTransactionId);
```
Elated to see inplace update go away :-)

asimrp replied:
Can we use `frozen_heap_inplace_delete()` instead?  I think this function is more appropriate because it does not incur the overhead of `HeapTupleSatisfiesUpdate()`, which is irrelevant for tables containing only frozen tuples, aka persistent tables.

asimrp commented:
This is such an important work, github text box is not enough to accommodate my plus ones!  Very happy to see this effort being driven to completion.

ashwinstar commented:
On Sat, Oct 15, 2016 at 1:27 PM, Asim RP <notifications@github.com> wrote:

> *@asimrp* commented on this pull request.
> ------------------------------
>
> In src/backend/cdb/cdbpersistentstore.c
> <https://github.com/greenplum-db/gpdb/pull/1210>:
>
> > -
> -	storeSharedData->freeTid = *persistentTid;
> -
> -	PersistentStore_FormTupleSetOurs(
> -							storeData,
> -							persistentRel->rd_att,
> -							freeValues,
> -							storeSharedData->maxFreeOrderNum,
> -							&prevFreeTid,
> -							&persistentTuple);
> -
> -	persistentTuple->t_self = *persistentTid;
> -		
> -	frozen_heap_inplace_update(persistentRel, persistentTuple);
> -
> +	simple_heap_delete_xid(persistentRel, persistentTid, FrozenTransactionId);
>
> Can we use frozen_heap_inplace_delete() instead? I think this function is
> more appropriate because it does not incur the overhead of
> HeapTupleSatisfiesUpdate(), which is irrelevant for tables containing
> only frozen tuples, aka persistent tables.
>
In my first iteration I had used frozen_heap_inplace_delete function. But
frankly I find it very funny to have a function named and doing
inplace_delete :-)  I was on the verge to delete that function but decided
to do it as separate patch. There is only one caller to it currently, so
don't even think its much tested as well.
Main point of not using that function, it takes tuple as input and not TID,
which means for PT FreeTuple case unnecessarily need to read the tuple
first to feed the function to delete, so that itself adds extra cost. So,
sure currently has extra cost of HeapTupleSatisfiesUpdate(), but more so
can then be argued for using heap_fetch to read tuple for PT. Instead
should be using specialized function as all inserts are frozen inserts
only. My current thinking is need to get away from as much specialized code
as possible and use generic routines if they can serve the purpose, for
maintenance and long term benefits. Plus, deletes to PT all come from DDL
statements only.

ashwinstar commented:
@hlinnaka Please can you have a look as well would be helpful to know if seems fine.

asimrp commented:
I understand now, why `frozen_heap_inplace_delete()` is not appropriate.  It is sensible not to use specialized interface for persistent tables.  And yeah, "inplace_delete" is a funny name!

hlinnaka commented:
I skimmed through this, but I don't understand the persistent table stuff very well, so I can't really analyze whether the new code is correct. But I like this a lot:

>  31 files changed, 237 insertions(+), 1943 deletions(-)

:-)

You can now also remove all the previous_free_tid fields from the gp_persistent_* system tables, right? The validate_previous_free_tid GUC can also go.

ashwinstar commented:
On Mon, Oct 17, 2016 at 12:21 AM, Heikki Linnakangas <
notifications@github.com> wrote:
>
> You can now also remove all the previous_free_tid fields from the
> gp_persistent_* system tables, right? The validate_previous_free_tid GUC
> can also go.
>
Yes added removal of previous_free_tid field as separate commit now. Also,
addressed other review comments from Asim and removed related unused gucs.

amilkh commented on src/backend/cdb/cdbpersistentfilesysobj.c:
```diff
@@ -1740,7 +1641,9 @@ PersistentFileSysObj_RemoveSegment(PersistentFileSysObjName *fsObjName,
 	READTUPLE_FOR_UPDATE_ERRCONTEXT_POP;
 
-	GpPersistent_GetCommonValues(
+	if (tupleCopy != NULL)
```
Is the addition of this `if (tupleCopy != NULL)` intentional?

ashwinstar replied:
Yes absolutely. We need to check now if tuple was found or not before trying to extract values from the same.

amilkh commented on src/backend/cdb/cdbpersistentstore.c:
```diff
@@ -1327,208 +682,45 @@ void PersistentStore_ReadTuple(
 	tuple.t_self = *readTid;
 
-	if (!heap_fetch(persistentRel, SnapshotAny,
+	if (heap_fetch(persistentRel, SnapshotAny,
```
is changing `heap_fetch` to `!heap_fetch` intentional here?

ashwinstar replied:
Yes absolutely.

amilkh commented:
This is a bit over my head, but how will this impact customers who want to upgrade GPDB to a newer version with these changes?

Also, are we testing this with anything other than ICG? Pretty big change set.

ashwinstar commented:
On Mon, Oct 17, 2016 at 5:56 PM, Amil Khanzada <notifications@github.com>
wrote:

> This is a bit over my head, but how will this impact customers who want to
> upgrade GPDB to a newer version with these changes?
>
This is again something needs to be handled as part of pg_upgrade.


> Also, are we testing this with anything other than ICG? Pretty big change
> set.
>
This is fully tested change-set from 4.3_STABLE.

amilkh commented:
Great!
@Eulerizeit for visibility as his team is working on the 4.3->5.0 upgrade tool.

Eulerizeit commented:
Thank you for the @ reply.
